File: lib-database-entities/src/main/java/com/ustadmobile/lib/db/entities/SelQuestion.java
package com.ustadmobile.lib.db.entities;

import com.ustadmobile.lib.database.annotation.UmEntity;
import com.ustadmobile.lib.database.annotation.UmPrimaryKey;
import com.ustadmobile.lib.database.annotation.UmSyncLastChangedBy;
import com.ustadmobile.lib.database.annotation.UmSyncLocalChangeSeqNum;
import com.ustadmobile.lib.database.annotation.UmSyncMasterChangeSeqNum;

/**
 * Represents the question within a question set.
 * eg: "Select the students who sit alone"
 *
 */
@UmEntity(tableId = 22)
public class SelQuestion {

    @UmPrimaryKey(autoGenerateSyncable = true)
    private long selQuestionUid;

    private String questionText;

    // -> SelQuestionSet - what set is this question a part of
    private long selQuestionSelQuestionSetUid;

    //The order.
    private int questionIndex;

    //If this question is to be assigned to all classes. (if not - not handled / implemented yet).
    boolean assignToAllClasses;

    //If this question allows for multiple nominations.
    boolean multiNominations;

    private int questionType;

    boolean questionActive;

    public boolean isQuestionActive() {
        return questionActive;
    }

    public void setQuestionActive(boolean questionActive) {
        this.questionActive = questionActive;
    }

    public int getQuestionType() {
        return questionType;
    }

    public void setQuestionType(int questionType) {
        this.questionType = questionType;
    }

    @UmSyncMasterChangeSeqNum
    private long selQuestionMasterChangeSeqNum;

    @UmSyncLocalChangeSeqNum
    private long selQuestionLocalChangeSeqNum;

    @UmSyncLastChangedBy
    private int selQuestionLastChangedBy;


    public boolean isAssignToAllClasses() {
        return assignToAllClasses;
    }

    public void setAssignToAllClasses(boolean assignToAllClasses) {
        this.assignToAllClasses = assignToAllClasses;
    }

    public boolean isMultiNominations() {
        return multiNominations;
    }

    public void setMultiNominations(boolean multiNominations) {
        this.multiNominations = multiNominations;
    }

    public String getQuestionText() {
        return questionText;
    }

    public void setQuestionText(String questionText) {
        this.questionText = questionText;
    }


    public int getQuestionIndex() {
        return questionIndex;
    }

    public void setQuestionIndex(int questionIndex) {
        this.questionIndex = questionIndex;
    }


    public int getSelQuestionLastChangedBy() {
        return selQuestionLastChangedBy;
    }

    public void setSelQuestionLastChangedBy(int selQuestionLastChangedBy) {
        this.selQuestionLastChangedBy = selQuestionLastChangedBy;
    }

    public long getSelQuestionLocalChangeSeqNum() {
        return selQuestionLocalChangeSeqNum;
    }

    public void setSelQuestionLocalChangeSeqNum(long selQuestionLocalChangeSeqNum) {
        this.selQuestionLocalChangeSeqNum = selQuestionLocalChangeSeqNum;
    }

    public long getSelQuestionMasterChangeSeqNum() {
        return selQuestionMasterChangeSeqNum;
    }

    public void setSelQuestionMasterChangeSeqNum(long selQuestionMasterChangeSeqNum) {
        this.selQuestionMasterChangeSeqNum = selQuestionMasterChangeSeqNum;
    }

    public long getSelQuestionSelQuestionSetUid() {
        return selQuestionSelQuestionSetUid;
    }

    public void setSelQuestionSelQuestionSetUid(long selQuestionSelQuestionSetUid) {
        this.selQuestionSelQuestionSetUid = selQuestionSelQuestionSetUid;
    }

    public long getSelQuestionUid() {
        return selQuestionUid;
    }

    public void setSelQuestionUid(long selQuestionUid) {
        this.selQuestionUid = selQuestionUid;
    }
}

File: lib-database-entities/src/commonMain/kotlin/com/ustadmobile/lib/db/entities/SelQuestion.kt
package com.ustadmobile.lib.db.entities

import androidx.room.Entity
import androidx.room.PrimaryKey
import com.ustadmobile.door.annotation.LastChangedBy
import com.ustadmobile.door.annotation.LocalChangeSeqNum
import com.ustadmobile.door.annotation.MasterChangeSeqNum
import com.ustadmobile.door.annotation.SyncableEntity
import kotlinx.serialization.Serializable

/**
 * Represents the question within a question set.
 * eg: "Select the students who sit alone"
 *
 */
@SyncableEntity(tableId = 22)
@Entity
@Serializable
class SelQuestion {

    @PrimaryKey(autoGenerate = true)
    var selQuestionUid: Long = 0

    var questionText: String? = null

    // -> SelQuestionSet - what set is this question a part of
    var selQuestionSelQuestionSetUid: Long = 0

    //The order.
    var questionIndex: Int = 0

    //If this question is to be assigned to all classes. (if not - not handled / implemented yet).
    var assignToAllClasses: Boolean = false

    //If this question allows for multiple nominations.
    var multiNominations: Boolean = false

    var questionType: Int = 0

    var questionActive: Boolean = false

    @MasterChangeSeqNum
    var selQuestionMasterChangeSeqNum: Long = 0

    @LocalChangeSeqNum
    var selQuestionLocalChangeSeqNum: Long = 0

    @LastChangedBy
    var selQuestionLastChangedBy: Int = 0

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || this::class != other::class) return false

        other as SelQuestion

        if (selQuestionUid != other.selQuestionUid) return false
        if (questionText != other.questionText) return false
        if (selQuestionSelQuestionSetUid != other.selQuestionSelQuestionSetUid) return false
        if (questionIndex != other.questionIndex) return false
        if (assignToAllClasses != other.assignToAllClasses) return false
        if (multiNominations != other.multiNominations) return false
        if (questionType != other.questionType) return false
        if (questionActive != other.questionActive) return false
        if (selQuestionMasterChangeSeqNum != other.selQuestionMasterChangeSeqNum) return false
        if (selQuestionLocalChangeSeqNum != other.selQuestionLocalChangeSeqNum) return false
        if (selQuestionLastChangedBy != other.selQuestionLastChangedBy) return false

        return true
    }

    override fun hashCode(): Int {
        var result = selQuestionUid.hashCode()
        result = 31 * result + (questionText?.hashCode() ?: 0)
        result = 31 * result + selQuestionSelQuestionSetUid.hashCode()
        result = 31 * result + questionIndex
        result = 31 * result + assignToAllClasses.hashCode()
        result = 31 * result + multiNominations.hashCode()
        result = 31 * result + questionType
        result = 31 * result + questionActive.hashCode()
        result = 31 * result + selQuestionMasterChangeSeqNum.hashCode()
        result = 31 * result + selQuestionLocalChangeSeqNum.hashCode()
        result = 31 * result + selQuestionLastChangedBy
        return result
    }


}
